x-pack/plugins/actions/server/lib/gen_ai_token_tracking.ts

```diff
@@ -7,17 +7,36 @@
 
 import { PassThrough, Readable } from 'stream';
 import { Logger } from '@kbn/logging';
+import { finished } from 'stream/promises';
 import { getTokenCountFromBedrockInvoke } from './get_token_count_from_bedrock_invoke';
 import { ActionTypeExecutorRawResult } from '../../common';
-import { getTokenCountFromOpenAIStream } from './get_token_count_from_openai_stream';
+import { getTokenCountFromOpenAI } from './get_token_count_from_openai_stream';
 import { getTokenCountFromInvokeStream, InvokeBody } from './get_token_count_from_invoke_stream';
 
-interface OwnProps {
-  actionTypeId: string;
-  logger: Logger;
-  result: ActionTypeExecutorRawResult<unknown>;
-  validatedParams: Record<string, unknown>;
-}
+type ValidatedParams =
+  | {
+      subAction: 'invokeAI';
+      subActionParams: {
+        messages: Array<{ content: string }>;
+      };
+    }
+  | {
+      subAction: 'run' | 'test';
+      subActionParams: {
+        body: string;
+      };
+    }
+  | {
+      subAction: 'invokeStream';
+      subActionParams: InvokeBody;
+    }
+  | {
+      subAction: 'invokeOpenAiStream';
+      subActionParams: { body: string };
+    };
```

Comment:
This makes the relationship between `subAction` and the `subActionParams` clear up front. This feels a little safer. Still, the type safety in this area is not great and most things just rely on various form of type casting.